## Упражнения по библиотеке Numpy

In [2]:
import numpy as np

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [ ]:
array = np.random.randint(10, size=20)
array

In [3]:
np.where(array<3, array, np.where(array>8, array, -1*array))

array([-4, -6,  1,  9,  9, -3, -4,  2, -5, -5, -6, -4, -7, -7,  2,  1, -5,
       -3,  0,  9])

**2.** Заменить максимальный элемент случайного массива на 0

In [5]:
max_val = np.amax(array)
np.where(array==max_val, 0, array)

array([4, 6, 1, 0, 0, 3, 4, 2, 5, 5, 6, 4, 7, 7, 2, 1, 5, 3, 0, 0])

**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [14]:
array2 = np.random.randint(10, size=5)
array2

array([0, 3, 5, 0, 8])

In [15]:
np.outer(array,array2)

array([[ 0, 12, 20,  0, 32],
       [ 0, 18, 30,  0, 48],
       [ 0,  3,  5,  0,  8],
       [ 0, 27, 45,  0, 72],
       [ 0, 27, 45,  0, 72],
       [ 0,  9, 15,  0, 24],
       [ 0, 12, 20,  0, 32],
       [ 0,  6, 10,  0, 16],
       [ 0, 15, 25,  0, 40],
       [ 0, 15, 25,  0, 40],
       [ 0, 18, 30,  0, 48],
       [ 0, 12, 20,  0, 32],
       [ 0, 21, 35,  0, 56],
       [ 0, 21, 35,  0, 56],
       [ 0,  6, 10,  0, 16],
       [ 0,  3,  5,  0,  8],
       [ 0, 15, 25,  0, 40],
       [ 0,  9, 15,  0, 24],
       [ 0,  0,  0,  0,  0],
       [ 0, 27, 45,  0, 72]])

In [18]:
# or
np.outer(array2,array)

array([[ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
         0,  0,  0,  0],
       [12, 18,  3, 27, 27,  9, 12,  6, 15, 15, 18, 12, 21, 21,  6,  3,
        15,  9,  0, 27],
       [20, 30,  5, 45, 45, 15, 20, 10, 25, 25, 30, 20, 35, 35, 10,  5,
        25, 15,  0, 45],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
         0,  0,  0,  0],
       [32, 48,  8, 72, 72, 24, 32, 16, 40, 40, 48, 32, 56, 56, 16,  8,
        40, 24,  0, 72]])

**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [150]:
A = np.random.randint(0, 5, size=(8, 3))
B = np.random.randint(0, 5, size=(2, 2))
print('Array A:')
print(A)
print('\nArray B:')
print(B)

temp = (A[..., np.newaxis, np.newaxis] == B)
rows = (temp.sum(axis=(1,2,3)) >= B.shape[1]).nonzero()[0]
print("\nRows of the array A that contain elements of each row of the array B:")
print(rows)

Array A:
[[1 0 3]
 [1 1 4]
 [2 4 0]
 [2 4 4]
 [0 3 4]
 [0 4 2]
 [2 4 0]
 [4 4 4]]

Array B:
[[0 2]
 [3 3]]

Rows of the array A that contain elements of each row of the array B:
[0 2 4 5 6]


**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [5]:
C = np.random.randint(0, 2, size=(10, 3))
print(C)

D = np.logical_and.reduce(C[:,1:] == C[:,:-1], axis = 1)
E = C[~D]

print('\nRows with unequal values:')
print(E)

'''
print('\nRows with unequal values:')
for i in range(10):
    if np.any(C[i]!=C[i][0]):
        print(i, C[i])
'''



[[0 1 0]
 [1 1 0]
 [0 0 1]
 [1 0 1]
 [1 0 0]
 [1 0 1]
 [1 1 1]
 [0 0 0]
 [1 1 1]
 [0 0 1]]

Rows with unequal values:
[[0 1 0]
 [1 1 0]
 [0 0 1]
 [1 0 1]
 [1 0 0]
 [1 0 1]
 [0 0 1]]


"\nprint('\nRows with unequal values:')\nfor i in range(10):\n    if np.any(C[i]!=C[i][0]):\n        print(i, C[i])\n"

**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [143]:
np.unique(A, axis=0)

array([[0, 1, 2],
       [0, 2, 2],
       [0, 2, 3],
       [1, 4, 0],
       [2, 4, 4],
       [3, 2, 3],
       [3, 3, 1],
       [4, 4, 3]])

______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [173]:
D = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]])
print(D)

[[1 0 1]
 [2 0 2]
 [3 0 3]
 [4 4 4]]


In [174]:
a = 1
for i in range(3):
    for j in range(3):
        if i == j and D[i][j] !=0:
            a *= D[i][j]
print(a)

3


In [176]:
diag = D.diagonal()
np.prod(diag, where = diag != 0)

3

* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [195]:
x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 2])

x.sort() == y.sort()

True

In [200]:
x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 2])

np.array_equal(np.sort(x), np.sort(y))

True

* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [233]:
x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])

filtered = []
j=0

for i in x:
    j += 1
    if i == 0 and j <= len(x)-1:
        filtered.append(x[j])

max(filtered)

5

In [269]:
x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])

ind = np.array(np.where(x == 0)) + 1
ind_limited = ind[ind <= len(x)-1]

max(np.take(x, ind_limited))

5

* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [309]:
x = np.array([2, 2, 2, 3, 3, 3, 5])
  
compressed = []
counts = []
count = 1
num = x[0]
for i in range(1,len(x)):
    if x[i] == num:
      count = count + 1
    else :
      compressed.append(num)
      counts.append(count)
      num = x[i]
      count = 1
compressed.append(num)
counts.append(count)
        
np.array(compressed), np.array(counts)

(array([2, 3, 5]), array([3, 3, 1]))

In [324]:
x = np.array([2, 2, 2, 3, 3, 3, 5])
  
compressed = []
counts = []
change = (x[1:] != x[:-1])
ind = np.append(np.where(change), len(x)-1)
count = np.diff(np.append(-1,ind))
 
for i in range(0,len(count)):
    compressed = np.append(compressed, x[ind][i])
    counts = np.append(counts, count[i])
 
np.array(compressed).astype(int), np.array(counts).astype(int)

(array([2, 3, 5]), array([3, 3, 1]))

* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [354]:
import scipy
import time

X = np.random.randint(0, 5, size=(5, 5))
Y = np.random.randint(0, 5, size=(5, 5))
print('Array X:')
print(X)
print('\nArray Y:')
print(Y)

Array X:
[[3 1 1 4 2]
 [2 1 0 4 0]
 [0 2 4 4 0]
 [3 3 2 4 1]
 [1 3 2 2 0]]

Array Y:
[[4 1 2 0 3]
 [0 1 3 1 4]
 [1 4 4 3 0]
 [1 0 1 2 4]
 [3 1 1 2 4]]


In [356]:
start_time = time.time()

P = np.add.outer(np.sum(X**2, axis=1), np.sum(Y**2, axis=1))
N = np.dot(X, Y.T)
print('\nResult:')
print(np.sqrt(P - 2*N))

print("--- %s seconds ---" % (time.time() - start_time))


Result:
[[4.35889894 5.09901951 5.19615242 3.60555128 2.82842712]
 [5.74456265 6.164414   5.19615242 4.79583152 4.69041576]
 [6.78232998 5.19615242 2.44948974 5.83095189 6.244998  ]
 [5.         5.65685425 3.31662479 5.19615242 4.24264069]
 [5.09901951 4.79583152 2.44948974 5.09901951 5.        ]]
--- 0.0020020008087158203 seconds ---


In [355]:
start_time = time.time()

print(scipy.spatial.distance.cdist(X, Y))

print("--- %s seconds ---" % (time.time() - start_time))

[[4.35889894 5.09901951 5.19615242 3.60555128 2.82842712]
 [5.74456265 6.164414   5.19615242 4.79583152 4.69041576]
 [6.78232998 5.19615242 2.44948974 5.83095189 6.244998  ]
 [5.         5.65685425 3.31662479 5.19615242 4.24264069]
 [5.09901951 4.79583152 2.44948974 5.09901951 5.        ]]
--- 0.0009870529174804688 seconds ---


_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [1]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [2]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [7]:
average_calories = calorie_stats.mean()
calories_difference = average_calories - 60

print('Average calories of competitors : ', average_calories)
print('Calories difference between CrunchieMunchies and competitors: ', calories_difference)

Average calories of competitors :  106.88311688311688
Calories difference between CrunchieMunchies and competitors:  46.883116883116884


3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [8]:
calorie_stats_sorted = np.sort(calorie_stats)

print(calorie_stats_sorted)

[ 50.  50.  50.  70.  70.  80.  90.  90.  90.  90.  90.  90.  90. 100.
 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100.
 100. 100. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 120. 120. 120. 120. 120. 120. 120. 120. 120. 120. 130.
 130. 140. 140. 140. 150. 150. 160.]


4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [ ]:
median_calories = np.median(calorie_stats)

print('Median: ', median_calories)
print('Average: ', average_calories)

5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [27]:
for nth_percentile in range (100):
    p = np.percentile(calorie_stats, nth_percentile)
    if p > 60:
        break

print(nth_percentile)

4


6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [33]:
size_all = np.size(calorie_stats)
size_more = np.size(calorie_stats[np.where(calorie_stats > 60)])

more_calories = 100*size_more/size_all

print(more_calories)

96.1038961038961


7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [34]:
calorie_std = np.std(calorie_stats)
print(calorie_std)

19.35718533390827


8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

As we can see, over 96% of other cereals have more calories than CrunchieMunchies. The average value of this indicator for them is 107. At the same time, the standard deviation is 19. That is, most competitors have calories from 88 to 126.
Thus, we can use in our advertising campaign the fact that CrunchieMunchies is a healthier product, as it contains significantly fewer calories than most other cereals.